### PR TITLE
fix: unicode characters in stderr on certain systems 

### LINKED ISSUE #1393: python 2 bytes/str issue
in `CommandLine` interface it `terminal_output` is `allatonce`, we do the following:

https://github.com/nipy/nipype/blob/master/nipype/interfaces/base.py#L1257

however on python 2, when the underlying command generates an error the returned string (at least on centos 7), contains some unsavory bytes. this results in an error as seen below 

one solution is to detect whether we are on python 2 and not decode the string.  any ideas how to solve this for both platforms simultaneously? @matthew-brett @bcipolli 

```
In [45]: from subprocess import Popen, PIPE

In [46]: proc = Popen('cp foo goo', shell=True, stdout=PIPE, stderr=PIPE)

In [47]: stdout, stderr = proc.communicate()

In [48]: stderr
Out[48]: 'cp: cannot stat \xe2\x80\x98foo\xe2\x80\x99: No such file or directory\n'

In [49]: str(stderr)
Out[49]: 'cp: cannot stat \xe2\x80\x98foo\xe2\x80\x99: No such file or directory\n'

In [50]: stderr.decode()
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-50-80fc7c733b30> in <module>()
----> 1 stderr.decode()

UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 16: ordinal not in range(128)

In [51]: stderr.decode('ISO-8859-1')
Out[51]: u'cp: cannot stat \xe2\x80\x98foo\xe2\x80\x99: No such file or directory\n'

In [53]: str(stderr.decode('ISO-8859-1'))
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
<ipython-input-53-e11a9dfd7944> in <module>()
----> 1 str(stderr.decode('ISO-8859-1'))

UnicodeEncodeError: 'ascii' codec can't encode characters in position 16-18: ordinal not in range(128)
```

### PR DESCRIPTION
**Pull request by satra**
- hopefully fixes #1393

### CODE REVIEW

File: nipype/interfaces/tests/test_base.py [hunk 1]
```diff
@@ -655,17 +655,17 @@ def test_CommandLine_output():
     ci = nib.CommandLine(command='ls -l')
     ci.inputs.terminal_output = 'allatonce'
     res = ci.run()
-    yield assert_equal, res.runtime.merged, ''
+    yield assert_equal, text_type(res.runtime.merged), text_type('')
```
Comment (matthew-brett) [inline nipype/interfaces/tests/test_base.py]: This should work fine without `text_type`.  On Python 2:

```
In [78]: unicode('hi there') == 'hi there'
Out[78]: True
```

On Python 3, literal strings are unicode, so you're comparing unicode to unicode already.

File: nipype/interfaces/tests/test_base.py [hunk 2]
```diff
@@ -655,17 +655,17 @@ def test_CommandLine_output():
...
     yield assert_true, name in res.runtime.stdout
     ci = nib.CommandLine(command='ls -l')
     ci.inputs.terminal_output = 'file'
     res = ci.run()
     yield assert_true, 'stdout.nipype' in res.runtime.stdout
-    yield assert_equal, type(res.runtime.stdout), type('hi')
+    yield assert_equal, type(text_type(res.runtime.stdout)), type(text_type('hi'))
```
Comment (matthew-brett) [inline nipype/interfaces/tests/test_base.py]: Maybe:

```
yield assert_equal, type(res.runtime.stdout), text_type
```

File: nipype/interfaces/base.py [hunk 3]
```diff
@@ -1258,17 +1259,17 @@ def _process(drain=0):
...
         if stderr and isinstance(stderr, bytes):
             try:
-                stderr = stderr.decode()
+                stderr = stderr.decode(locale.getdefaultlocale()[1])
             except UnicodeDecodeError:
                 stderr = stderr.decode("ISO-8859-1")
 
-        result['stdout'] = str(stdout).split('\n')
-        result['stderr'] = str(stderr).split('\n')
+        result['stdout'] = text_type(stdout).split(text_type('\n'))
```
Comment (matthew-brett) [inline nipype/interfaces/base.py]: So - I think should should be:

```
result['stdout'] = stdout.split('\n')
```

Same idea for the other uses of `text_type`.

**Comment by satra** [comment]
thank you @matthew-brett - i think i have cleaned this up for this issue.

**Comment by PeerHerholz** [comment]
Hi,

I stumbled across a weird error (at least for me) when running my nipype script and I think
it might be related to the error discussed here. 

Here's the error message: 

Exception in thread Thread-7:
Traceback (most recent call last):
  File "/Users/peerherholz/anaconda/envs/py27/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/Users/peerherholz/anaconda/envs/py27/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/peerherholz/anaconda/envs/py27/lib/python2.7/multiprocessing/pool.py", line 389, in _handle_results
    task = get()
TypeError: ('function takes exactly 5 arguments (7 given)', <type 'exceptions.UnicodeDecodeError'>, ('utf8', 'MATLAB code threw an exception:\nNo executable modules, but still unresolved dependencies or incomplete module inputs.\nFile:/Applications/spm12/spm_jobman.m\nName:/Applications/spm12/spm_jobman.m\nLine:47\nFile:Path\to\my\file\realign\pyscript_realign.m\nName:fill_run_job\nLine:115\nFile:pm_jobman\nName:pyscript_realign\nLine:461\nFile:\xf7\nName:\x1a\nLine:', 455, 456, 'invalid start byte', '\nInterface MatlabCommand failed to run.', '\nInterface Realign failed to run.'))

It only appears when I merge the 3D .nii files into a 4D file using Fsl Merge before realignment.  Using 3D files as input for the realignment node works fine.

@satra: I checked if the fixes you made are included in my version and they are. I also updated nipype and it's dependencies. Unfortunately, it didn't resolve the error. 

I'm using python 2.7..12 in OS X El Capitan, both, in a VM and on my macbook. 
In either case, the error appears. 

Can you think of anything that I could try to fix that?

**Comment by satra** [comment]
@PeerHerholz - there are two pieces to this error:

1. the utf output from matlab when it errors that should be parsed and decoded properly 
2. the issue with realign itself. are you by any chance giving spm a compressed nifti file? can you ensure the outputtype of fslmerge is NIFTI and not NIFIT_GZ

**Comment by PeerHerholz** [comment]
@satra thank you very much for your fast response and the pointers. 

I indeed forgot to set the output type in my merge node. After setting it to NIFTI the realignment and coregister nodes work fine (I'm basically trying to run the "classic" SPM pipeline, including unified segmentation). However, now the same error appears either in the segmentation or normalize _structural node (weirdly, depending on the machine I us). 
As far as I can tell and based on the detailed pipeline graph, the flow of files fits. 

I thought it might be related to the default encoding, as I'm using Anaconda and Spyder, but no matter how I call the script (e.g. within spyder, command prompt, etc.) the error remains the same. 

Sorry, if this is a very noobish thing to guess / ask.

**Comment by satra** [comment]
@PeerHerholz - let me take a look

**Comment by PeerHerholz** [comment]
@satra thank you very much.

After some additional tests, I can provide you with some (hopefully) helpful information:

After I changed the my pipeline from multiproc to serial the node seem to work, but only for one 
run. After that the following error appears: 

170214-20:09:40,751 workflow INFO:
         Traceback (most recent call last):
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/pipeline/plugins/linear.py", line 43, in run
    node.run(updatehash=updatehash)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/pipeline/engine/nodes.py", line 366, in run
    self._run_interface()
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/pipeline/engine/nodes.py", line 476, in _run_interface
    self._result = self._run_command(execute)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/pipeline/engine/nodes.py", line 607, in _run_command
    result = self._interface.run()
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/base.py", line 1085, in run
    runtime = self._run_wrapper(runtime)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/base.py", line 1033, in _run_wrapper
    runtime = self._run_interface(runtime)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/spm/base.py", line 323, in _run_interface
    results = self.mlab.run()
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/base.py", line 1085, in run
    runtime = self._run_wrapper(runtime)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/base.py", line 1728, in _run_wrapper
    runtime = self._run_interface(runtime)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/matlab.py", line 144, in _run_interface
    runtime = super(MatlabCommand, self)._run_interface(runtime)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/base.py", line 1759, in _run_interface
    redirect_x=self._redirect_x)
  File "/Users/peerherholz/Downloads/nipype-0.13.0-rc1/nipype/interfaces/base.py", line 1514, in run_command
    stderr = stderr.decode(default_encoding)
  File "/Users/peerherholz/anaconda/lib/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xf7 in position 440: invalid start byte

After I upgraded nipype to the developer version, the same thing happened. 
Also changing the apparently crucial line in base.py from 
stderr = stderr.decode(default_encoding) to
stderr = stderr.decode(locale.getdefaultlocale()[1]) as in #1394 
had no effect. The same error indicating the same lines of code appears. 

I'm truly sorry for the hustle.

**Pull request merged**